src/main/stream_query_result.cpp

```diff
@@ -3,26 +3,34 @@
 #include "duckdb/main/client_context.hpp"
 #include "duckdb/main/materialized_query_result.hpp"
 #include "duckdb/common/box_renderer.hpp"
+#include "duckdb/main/database.hpp"
 
 namespace duckdb {
 
 StreamQueryResult::StreamQueryResult(StatementType statement_type, StatementProperties properties,
-                                     shared_ptr<ClientContext> context_p, vector<LogicalType> types,
-                                     vector<string> names)
+                                     vector<LogicalType> types, vector<string> names,
+                                     ClientProperties client_properties, shared_ptr<BufferedData> data)
     : QueryResult(QueryResultType::STREAM_RESULT, statement_type, std::move(properties), std::move(types),
-                  std::move(names), context_p->GetClientProperties()),
-      context(std::move(context_p)) {
-	D_ASSERT(context);
+                  std::move(names), std::move(client_properties)),
+      buffered_data(std::move(data)) {
+	context = buffered_data->GetContext();
 }
 
 StreamQueryResult::~StreamQueryResult() {
+	if (!context) {
+		return;
+	}
+	auto lock = context->LockContext();
+	if (IsOpenInternal(*lock)) {
```

Comment:
It seems that it is, without it we are leaking objects, can't seem to find out the root cause of this
It might be related to the ClientData, it looks like it's not being cleaned up

Reply:
Maybe we're somehow creating a reference cycle?
On mac when I breakpoint `ClientContext::~ClientContext` it's hit twice, once when the `Database db` is created, and another time when the `shared_ptr<ClientContext>` hits refcount 0.

On Linux when I breakpoint that same symbol, it's only hit once, when the `Database db` is created.
Which sounds to me like the refcount never hits 0

Reply:
Maybe it has something to do with storing the `shared_ptr<ClientContext>` inside the `BufferedData` object.

The ClientContext holds the ActiveQueryContext which holds the Executor, the Executor holds the PhysicalBufferedCollector, which has a `shared_ptr<BufferedData>`

The StreamQueryResult also has a `shared_ptr<BufferedData>`

So the `shared_ptr<ClientContext>` refcount cant hit 0 because there are objects inside the ClientContext that still hold a shared_ptr to it?

Reply:
That makes sense, maybe the `shared_ptr<ClientContext>` in the `BufferedData` can be turned into a `weak_ptr` instead?

Reply:
I have turned the `shared_ptr`s in `BufferedData` and `BufferedCollectorGlobalState` into `weak_ptr`.

Now the only object that has an owning reference is the `StreamQueryResult`.
This is necessary because for normal execution we do the following:
1. run the pipelines to completion
2. get the result
3. cleanup the query
4. fetch from the result, which is stand-alone from the ClientContext at this point

Instead we do the following:
1. run until all pipelines are blocked
2. get the result
3. fetch until completion
4. cleanup the query.

If the result isnt fetched until completion, we will not clean up the query, letting the sink states stay alive inside the Executor, which is owned by the ClientContext.
If those sink states or objects they own have an owning `shared_ptr`, this will prevent destruction of the ClientContext and everything it owns.